src/main/java/beam/physsim/cch/RoutingFrameworkTravelTimeCalculator.scala

```diff
@@ -0,0 +1,230 @@
+                    .getOrElse(origin, getClosestVertexId(origin))
+                  val secondVertexId: Long = coord2VertexId
+                    .getOrElse(destination, getClosestVertexId(destination))
+
+                  odNumber = odNumber + 1
+
+                  OD(firstVertexId, secondVertexId)
+              }
+
+            val odStream = ods
+              .flatMap(od => Stream.range(0, odsFactor, 1).map(_ => od))
+            routingFrameworkWrapper.generateOd(iterationNumber, hour, odStream)
+
+            logger.info("Generated {} ods, for hour {} in {} ms", odNumber, hour, stopWatch.getTime)
+          }
+      }
+
+    Await.result(Future.sequence(odsCreationFutures), 10.minutes)
+
+    val hour2Way2TravelTimes = hour2Events.keys.toList.sorted.map { hour =>
+      val stopWatch: StopWatch = new StopWatch
+      stopWatch.start()
+
+      logger.info("Starting traffic assignment for hour {}", hour)
+      val way2TravelTime =
+        routingFrameworkWrapper.assignTrafficAndFetchWay2TravelTime(iterationNumber, hour)
+      logger.info("Traffic assignment for hour {} is finished in {} ms", hour, stopWatch.getTime)
+
+      (hour, way2TravelTime)
+    }.toMap
+
+    val totalNumberOfLinks: Int = links.size
+    var linksFailedToResolve = 0
+
+    val linkId2TravelTimeByHour = links
+      .groupBy(linkWayId)
+      .flatMap {
+        case (maybeWayId, linksInWay) =>
+          linksInWay.map { link =>
+            val travelTimeByHour = maybeWayId match {
+              case Some(wayId) =>
+                var speedFoundForAtLeastOneHour = false
+
+                val linkTravelTimeByHour = Array.tabulate(maxHour) { hour =>
+                  hour2Way2TravelTimes.get(hour).flatMap(_.get(wayId)) match {
+                    case Some(travelTime) =>
+                      speedFoundForAtLeastOneHour = true
+                      travelTime / linksInWay.size
+
+                    case None =>
+                      link.getLength / link.getFreespeed
+                  }
+                }
+
+                if (!speedFoundForAtLeastOneHour)
+                  linksFailedToResolve = linksFailedToResolve + 1
+
+                linkTravelTimeByHour
+
+              case None =>
+                linksFailedToResolve = linksFailedToResolve + 1
+                Array.fill(maxHour) { link.getLength / link.getFreespeed }
+            }
+            link.getId.toString -> travelTimeByHour
+          }
+      }
+
+    logger.info("Total links: {}, failed to assign travel time: {}", totalNumberOfLinks, linksFailedToResolve)
+
+    logger.info("Created travel times in {} ms", System.currentTimeMillis - startTime)
+
+    linkId2TravelTimeByHour.asJava
+  }
+
+  private def generateHour2Events(
+    pathTraversalEvents: util.Collection[PathTraversalEvent]
+  ): Map[Int, List[TravelInfo]] = {
+    val preliminaryHour2Events: Map[Int, List[PathTraversalEvent]] =
+      pathTraversalEvents.toStream
+        .map(x => x.departureTime / 3600 -> x)
```

Comment:
Another reason to create a Utils.merge or something? You could even put it in an implicit class on top of Lists or something?

Reply:
looks like in all cases it is just a simple group by =\, replaced